### ISSUE TITLE
When adding a listener to a child component I want to be able to listen to only the first detach event so that I can remove the listener

### ISSUE THREAD

**Issue by Artur-** [vaadin/flow#3357]
For example, a layout style component which adds a click listener to all its child components. The layout wants to remove the click listener when a child component is removed from the layout. The most logical place to do this is in a detach listener for the child, as it does not matter then in which way the child component has been detached. This seems a bit hard to do though as you would want to write something like
```
Registration registration = child.addClickListener(e -> { ... });
Registration detachRegistration = child.addDetachListener(e -> {
    registration.remove();
    detachRegistration.remove();
});
```
which does not work because you cannot refer to `detachRegistration` inside the detach listener:
 "The local variable detachRegistration may not have been initialized".

What I would really have liked to express is either:
1. "Add this click listener and automatically remove it on detach"
or
2. "Add this click listener and the first time the child component is detached, unregister the click listener"

**Comment by pleku** [comment]
What if the `event` would have a way for detaching that specific listener ? E.g.
```java
Registration detachRegistration = child.addDetachListener(e -> {
    registration.remove();
    event.removeCurrentListener(); // works only once per triggered listener, throws if called twice for triggered listener
});
``

**Comment by denis-anisimov** [comment]
From the API point of view the event doesn't know anything about listeners.
I think it's not a good way to achieve the requirement.
Instead a listener may have `boolean` as a return value type instead of `void`.
Once `true` is returned the listener is removed.

**Comment by pleku** [comment]
Forcing a return value would prevent anyone from having _oneliner_ listeners. That might end up being really annoying quickly. Also it would make it weird for trying to use the same listener in multiple places - which registrations does the returned value remove, all or just for that event ?

Another option is to make the `Listener` have state by adding link to the registration, e.g. then it would be possible to just call `remove()`. But that also is a bit weird when you have the same listener used in multiple places.

**Comment by Legioth** [comment]
One possibility that would help in this case would be a special overload of `addXyzListener` that is only fired once and then automatically removes itself. With this functionality, the example could be written in this way:

```
Registration registration = child.addClickListener(e -> { ... });
child.addDetachListener(e ->  registration.remove(), RegistrationMode.FIRE_ONCE);
```

**Comment by pleku** [comment]
So this could be handled by either
- the _fire only once_ way
- adding the registration to the event as proposed in https://github.com/vaadin/flow/issues/3357

We should do a quick DX testing to discover which one is preferred and make that one happen

**Comment by heruan** [comment]
I've seen the *fire only once* pattern in many frameworks, e.g.

- NodeJS: [`on`](https://nodejs.org/api/events.html#events_emitter_on_eventname_listener) vs [`once`](https://nodejs.org/api/events.html#events_emitter_once_eventname_listener)
- jQuery: [`on`](https://api.jquery.com/on/) vs [`one`](https://api.jquery.com/one/)

I see that there is #4073 to provide the listener a way to unregister itself, but it would be nice to have both ways to simplify e.g. use cases when you have the same listener and you want to register it once for some components and always for others.

**Comment by raphw** [comment]
Is this still a priority? I would benefit from this, too.

**Comment by Legioth** [comment]
A way of doing this was already implemented. The event class nowadays has an `unregisterListener()` method that you can call from inside the listener to prevent it from being fired again.

I'm not sure if we even need any dedicated API for registering listeners for single use when you can use `unregisterListener()` instead.

**Comment by raphw** [comment]
Alright, I overlooked that. Maybe this ticket can be closed then?

**Comment by Legioth** [comment]
Fixed by #4073.
